x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java

```diff
@@ -425,8 +425,12 @@ public void testRefreshingMultipleTimesWithinWindowSucceeds() throws Exception {
                     return;
                 }
                 threadSecurityClient.refreshToken(refreshRequest, ActionListener.wrap(result -> {
-                    accessTokens.add(result.getTokenString());
-                    refreshTokens.add(result.getRefreshToken());
+                    if (accessTokens.get().size() > 0 && accessTokens.get().contains(result.getTokenString()) == false ||
+                        refreshTokens.get().size() > 0 && refreshTokens.get().contains(result.getRefreshToken()) == false) {
+                        newToken.set(true);
+                    }
+                    accessTokens.get().add(result.getTokenString());
+                    refreshTokens.get().add(result.getRefreshToken());
```

Comment:
The root problem is that `HashSet` is not thread-safe. And wrapping it in an `AtomicReference` won't do any good. I recommend using `ConcurrentHashMap` and keeping the original logic - without `newToken` .

Reply:
Gotcha